File: test/src/main/java/cz/vhromada/common/test/facade/MovableParentFacadeTest.java
package cz.vhromada.common.test.facade;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.SoftAssertions.assertSoftly;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

import java.util.Collections;
import java.util.List;

import cz.vhromada.common.Movable;
import cz.vhromada.common.converter.MovableConverter;
import cz.vhromada.common.facade.MovableParentFacade;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.validator.MovableValidator;
import cz.vhromada.common.validator.ValidationType;
import cz.vhromada.validation.result.Event;
import cz.vhromada.validation.result.Result;
import cz.vhromada.validation.result.Severity;
import cz.vhromada.validation.result.Status;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

/**
 * An abstract class represents test for {@link MovableParentFacade}.
 *
 * @param <T> type of entity data
 * @param <U> type of domain data
 * @author Vladimir Hromada
 */
@ExtendWith(MockitoExtension.class)
@SuppressWarnings("JUnitTestMethodInProductSource")
public abstract class MovableParentFacadeTest<T extends Movable, U extends Movable> {

    /**
     * Result for invalid data
     */
    private static final Result<Void> INVALID_DATA_RESULT = Result.error("DATA_INVALID", "Data must be valid.");

    /**
     * Instance of {@link MovableService}
     */
    @Mock
    private MovableService<U> service;

    /**
     * Instance of {@link MovableConverter}
     */
    @Mock
    private MovableConverter<T, U> converter;

    /**
     * Instance of {@link MovableValidator}
     */
    @Mock
    private MovableValidator<T> validator;

    /**
     * Instance of {@link MovableParentFacade}
     */
    private MovableParentFacade<T> facade;

    /**
     * Initializes facade for movable data.
     */
    @BeforeEach
    public void setUp() {
        facade = getFacade();
    }

    /**
     * Test method for {@link MovableParentFacade#newData()}.
     */
    @Test
    void newData() {
        final Result<Void> result = facade.newData();

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        verify(service).newData();
        verifyNoMoreInteractions(service);
        verifyZeroInteractions(converter, validator);
    }

    /**
     * Test method for {@link MovableParentFacade#getAll()}.
     */
    @Test
    void getAll() {
        final List<U> domainList = List.of(newDomain(1), newDomain(2));
        final List<T> entityList = List.of(newEntity(1), newEntity(2));

        when(service.getAll()).thenReturn(domainList);
        when(converter.convertBack(anyList())).thenReturn(entityList);

        final Result<List<T>> result = facade.getAll();

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getData()).isEqualTo(entityList);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        verify(service).getAll();
        verify(converter).convertBack(domainList);
        verifyNoMoreInteractions(service, converter);
        verifyZeroInteractions(validator);
    }


    /**
     * Test method for {@link MovableParentFacade#get(Integer)} with existing data.
     */
    @Test
    void get_ExistingData() {
        final U domain = newDomain(1);
        final T entity = newEntity(1);

        when(service.get(any(Integer.class))).thenReturn(domain);
        when(converter.convertBack(any(getDomainClass()))).thenReturn(entity);

        final Result<T> result = facade.get(1);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getData()).isEqualTo(entity);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        verify(service).get(1);
        verify(converter).convertBack(domain);
        verifyNoMoreInteractions(service, converter);
        verifyZeroInteractions(validator);
    }

    /**
     * Test method for {@link MovableParentFacade#get(Integer)} with not existing data.
     */
    @Test
    void get_NotExistingData() {
        when(service.get(any(Integer.class))).thenReturn(null);

        final Result<T> result = facade.get(Integer.MAX_VALUE);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getData()).isNull();
            softly.assertThat(result.getEvents()).isEmpty();
        });

        verify(service).get(Integer.MAX_VALUE);
        verify(converter).convertBack((U) null);
        verifyNoMoreInteractions(service, converter);
        verifyZeroInteractions(validator);
    }

    /**
     * Test method for {@link MovableParentFacade#get(Integer)} with null data.
     */
    @Test
    void get_NullData() {
        final Result<T> result = facade.get(null);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getData()).isNull();
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "ID_NULL", "ID mustn't be null.")));
        });

        verifyZeroInteractions(service, converter, validator);
    }

    /**
     * Test method for {@link MovableParentFacade#add(Movable)}.
     */
    @Test
    void add() {
        final T entity = newEntity(null);
        final U domain = newDomain(null);

        when(converter.convert(any(getEntityClass()))).thenReturn(domain);
        when(validator.validate(any(getEntityClass()), any())).thenReturn(new Result<>());

        final Result<Void> result = facade.add(entity);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        verify(service).add(domain);
        verify(converter).convert(entity);
        verify(validator).validate(entity, ValidationType.NEW, ValidationType.DEEP);
        verifyNoMoreInteractions(service, converter, validator);
    }

    /**
     * Test method for {@link MovableParentFacade#add(Movable)} with invalid data.
     */
    @Test
    void add_InvalidData() {
        final T entity = newEntity(Integer.MAX_VALUE);

        when(validator.validate(any(getEntityClass()), any())).thenReturn(INVALID_DATA_RESULT);

        final Result<Void> result = facade.add(entity);

        assertThat(result).isEqualTo(INVALID_DATA_RESULT);

        verify(validator).validate(entity, ValidationType.NEW, ValidationType.DEEP);
        verifyNoMoreInteractions(validator);
        verifyZeroInteractions(service, converter);
    }

    /**
     * Test method for {@link MovableParentFacade#update(Movable)}.
     */
    @Test
    void update() {
        final T entity = newEntity(1);
        final U domain = newDomain(1);

        initUpdateMock(domain);

        final Result<Void> result = facade.update(entity);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        verifyUpdateMock(entity, domain);
        verifyNoMoreInteractions(service, converter, validator);
    }

    /**
     * Test method for {@link MovableParentFacade#update(Movable)} with invalid data.
     */
    @Test
    void update_InvalidData() {
        final T entity = newEntity(Integer.MAX_VALUE);

        when(validator.validate(any(getEntityClass()), any())).thenReturn(INVALID_DATA_RESULT);

        final Result<Void> result = facade.update(entity);

        assertThat(result).isEqualTo(INVALID_DATA_RESULT);

        verify(validator).validate(entity, ValidationType.UPDATE, ValidationType.EXISTS, ValidationType.DEEP);
        verifyNoMoreInteractions(validator);
        verifyZeroInteractions(service, converter);
    }

    /**
     * Test method for {@link MovableParentFacade#remove(Movable)}.
     */
    @Test
    void remove() {
        final T entity = newEntity(1);
        final U domain = newDomain(1);

        when(service.get(any(Integer.class))).thenReturn(domain);
        when(validator.validate(any(getEntityClass()), any())).thenReturn(new Result<>());

        final Result<Void> result = facade.remove(entity);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        verify(service).get(1);
        verify(service).remove(domain);
        verify(validator).validate(entity, ValidationType.EXISTS);
        verifyNoMoreInteractions(service, validator);
        verifyZeroInteractions(converter);
    }

    /**
     * Test method for {@link MovableParentFacade#remove(Movable)} with invalid data.
     */
    @Test
    void remove_InvalidData() {
        final T entity = newEntity(Integer.MAX_VALUE);

        when(validator.validate(any(getEntityClass()), any())).thenReturn(INVALID_DATA_RESULT);

        final Result<Void> result = facade.remove(entity);

        assertThat(result).isEqualTo(INVALID_DATA_RESULT);

        verify(validator).validate(entity, ValidationType.EXISTS);
        verifyNoMoreInteractions(validator);
        verifyZeroInteractions(service, converter);
    }

    /**
     * Test method for {@link MovableParentFacade#duplicate(Movable)}.
     */
    @Test
    void duplicate() {
        final T entity = newEntity(1);
        final U domain = newDomain(1);

        when(service.get(any(Integer.class))).thenReturn(domain);
        when(validator.validate(any(getEntityClass()), any())).thenReturn(new Result<>());

        final Result<Void> result = facade.duplicate(entity);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        verify(service).get(1);
        verify(service).duplicate(domain);
        verify(validator).validate(entity, ValidationType.EXISTS);
        verifyNoMoreInteractions(service, validator);
        verifyZeroInteractions(converter);
    }

    /**
     * Test method for {@link MovableParentFacade#duplicate(Movable)} with invalid data.
     */
    @Test
    void duplicate_InvalidData() {
        final T entity = newEntity(Integer.MAX_VALUE);

        when(validator.validate(any(getEntityClass()), any())).thenReturn(INVALID_DATA_RESULT);

        final Result<Void> result = facade.duplicate(entity);

        assertThat(result).isEqualTo(INVALID_DATA_RESULT);

        verify(validator).validate(entity, ValidationType.EXISTS);
        verifyNoMoreInteractions(validator);
        verifyZeroInteractions(service, converter);
    }

    /**
     * Test method for {@link MovableParentFacade#moveUp(Movable)}.
     */
    @Test
    void moveUp() {
        final T entity = newEntity(1);
        final U domain = newDomain(1);

        when(service.get(any(Integer.class))).thenReturn(domain);
        when(validator.validate(any(getEntityClass()), any())).thenReturn(new Result<>());

        final Result<Void> result = facade.moveUp(entity);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        verify(service).get(1);
        verify(service).moveUp(domain);
        verify(validator).validate(entity, ValidationType.EXISTS, ValidationType.UP);
        verifyNoMoreInteractions(service, validator);
        verifyZeroInteractions(converter);
    }

    /**
     * Test method for {@link MovableParentFacade#moveUp(Movable)} with invalid data.
     */
    @Test
    void moveUp_InvalidData() {
        final T entity = newEntity(Integer.MAX_VALUE);

        when(validator.validate(any(getEntityClass()), any())).thenReturn(INVALID_DATA_RESULT);

        final Result<Void> result = facade.moveUp(entity);

        assertThat(result).isEqualTo(INVALID_DATA_RESULT);

        verify(validator).validate(entity, ValidationType.EXISTS, ValidationType.UP);
        verifyNoMoreInteractions(validator);
        verifyZeroInteractions(service, converter);
    }

    /**
     * Test method for {@link MovableParentFacade#moveDown(Movable)}.
     */
    @Test
    void moveDown() {
        final T entity = newEntity(1);
        final U domain = newDomain(1);

        when(service.get(any(Integer.class))).thenReturn(domain);
        when(validator.validate(any(getEntityClass()), any())).thenReturn(new Result<>());

        final Result<Void> result = facade.moveDown(entity);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        verify(service).get(1);
        verify(service).moveDown(domain);
        verify(validator).validate(entity, ValidationType.EXISTS, ValidationType.DOWN);
        verifyNoMoreInteractions(service, validator);
        verifyZeroInteractions(converter);
    }

    /**
     * Test method for {@link MovableParentFacade#moveDown(Movable)} with invalid data.
     */
    @Test
    void moveDown_InvalidData() {
        final T entity = newEntity(Integer.MAX_VALUE);

        when(validator.validate(any(getEntityClass()), any())).thenReturn(INVALID_DATA_RESULT);

        final Result<Void> result = facade.moveDown(entity);

        assertThat(result).isEqualTo(INVALID_DATA_RESULT);

        verify(validator).validate(entity, ValidationType.EXISTS, ValidationType.DOWN);
        verifyNoMoreInteractions(validator);
        verifyZeroInteractions(service, converter);
    }

    /**
     * Test method for {@link MovableParentFacade#updatePositions()}.
     */
    @Test
    void updatePositions() {
        final Result<Void> result = facade.updatePositions();

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        verify(service).updatePositions();
        verifyNoMoreInteractions(service);
        verifyZeroInteractions(converter, validator);
    }

    /**
     * Returns service for movable data.
     *
     * @return service for movable data
     */
    protected MovableService<U> getService() {
        return service;
    }

    /**
     * Returns converter for movable data.
     *
     * @return converter for movable data
     */
    protected MovableConverter<T, U> getConverter() {
        return converter;
    }

    /**
     * Returns validator for movable data.
     *
     * @return validator for movable data
     */
    protected MovableValidator<T> getValidator() {
        return validator;
    }

    /**
     * Initializes mock for update.
     *
     * @param domain domain
     */
    protected void initUpdateMock(final U domain) {
        when(converter.convert(any(getEntityClass()))).thenReturn(domain);
        when(validator.validate(any(getEntityClass()), any())).thenReturn(new Result<>());
    }

    /**
     * Verifies mock for update.
     *
     * @param entity entity
     * @param domain domain
     */
    protected void verifyUpdateMock(final T entity, final U domain) {
        verify(service).update(domain);
        verify(converter).convert(entity);
        verify(validator).validate(entity, ValidationType.UPDATE, ValidationType.EXISTS, ValidationType.DEEP);
    }

    /**
     * Returns facade for movable data for parent data.
     *
     * @return facade for movable data for parent data
     */
    protected abstract MovableParentFacade<T> getFacade();

    /**
     * Returns entity.
     *
     * @param id ID
     * @return entity
     */
    protected abstract T newEntity(Integer id);

    /**
     * Returns domain.
     *
     * @param id ID
     * @return domain
     */
    protected abstract U newDomain(Integer id);

    /**
     * Returns entity class.
     *
     * @return entity class.
     */
    protected abstract Class<T> getEntityClass();

    /**
     * Returns domain class.
     *
     * @return domain class.
     */
    protected abstract Class<U> getDomainClass();

}

File: Test/src/main/kotlin/cz/vhromada/common/test/facade/MovableParentFacadeTest.kt
package cz.vhromada.common.test.facade

import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.verifyNoMoreInteractions
import com.nhaarman.mockitokotlin2.verifyZeroInteractions
import com.nhaarman.mockitokotlin2.whenever
import cz.vhromada.common.Movable
import cz.vhromada.common.facade.MovableParentFacade
import cz.vhromada.common.mapper.Mapper
import cz.vhromada.common.service.MovableService
import cz.vhromada.common.validator.MovableValidator
import cz.vhromada.common.validator.ValidationType
import cz.vhromada.validation.result.Result
import cz.vhromada.validation.result.Status
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions.assertSoftly
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension

/**
 * Result for invalid data
 */
private val INVALID_DATA_RESULT = Result.error<Unit>("DATA_INVALID", "Data must be valid.")

/**
 * An abstract class represents test for [MovableParentFacade].
 *
 * @param <T> type of entity data
 * @param <U> type of domain data
 * @author Vladimir Hromada
 */
@ExtendWith(MockitoExtension::class)
@Suppress("FunctionName")
abstract class MovableParentFacadeTest<T : Movable, U : Movable> {

    /**
     * Instance of [MovableService]
     */
    @Mock
    protected lateinit var service: MovableService<U>

    /**
     * Instance of [Mapper]
     */
    @Mock
    protected lateinit var mapper: Mapper<T, U>

    /**
     * Instance of [MovableValidator]
     */
    @Mock
    protected lateinit var validator: MovableValidator<T>

    /**
     * Instance of [MovableParentFacade]
     */
    private lateinit var facade: MovableParentFacade<T>

    /**
     * Initializes facade for movable data.
     */
    @BeforeEach
    open fun setUp() {
        facade = getFacade()
    }

    /**
     * Test method for [MovableParentFacade.newData].
     */
    @Test
    fun newData() {
        val result = facade.newData()

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.events()).isEmpty()
        }

        verify(service).newData()
        verifyNoMoreInteractions(service)
        verifyZeroInteractions(mapper, validator)
    }

    /**
     * Test method for [MovableParentFacade.getAll].
     */
    @Test
    fun getAll() {
        val domainList = listOf(newDomain(1), newDomain(2))
        val entityList = listOf(newEntity(1), newEntity(2))

        whenever(service.getAll()).thenReturn(domainList)
        whenever(mapper.mapBack(any<List<U>>())).thenReturn(entityList)

        val result = facade.getAll()

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.data).isEqualTo(entityList)
            it.assertThat(result.events()).isEmpty()
        }

        verify(service).getAll()
        verify(mapper).mapBack(domainList)
        verifyNoMoreInteractions(service, mapper)
        verifyZeroInteractions(validator)
    }


    /**
     * Test method for [MovableParentFacade.get] with existing data.
     */
    @Test
    fun get_ExistingData() {
        val domain = newDomain(1)
        val entity = newEntity(1)

        whenever(service.get(any())).thenReturn(domain)
        whenever(mapper.mapBack(anyDomain())).thenReturn(entity)

        val result = facade.get(1)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.data).isEqualTo(entity)
            it.assertThat(result.events()).isEmpty()
        }

        verify(service).get(1)
        verify(mapper).mapBack(domain)
        verifyNoMoreInteractions(service, mapper)
        verifyZeroInteractions(validator)
    }

    /**
     * Test method for [MovableParentFacade.get] with not existing data.
     */
    @Test
    fun get_NotExistingData() {
        whenever(service.get(any())).thenReturn(null)

        val result = facade.get(Integer.MAX_VALUE)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.data).isNull()
            it.assertThat(result.events()).isEmpty()
        }

        verify(service).get(Integer.MAX_VALUE)
        verifyNoMoreInteractions(service, mapper)
        verifyZeroInteractions(mapper, validator)
    }

    /**
     * Test method for [MovableParentFacade.add].
     */
    @Test
    fun add() {
        val entity = newEntity(null)
        val domain = newDomain(null)

        whenever(mapper.map(anyEntity())).thenReturn(domain)
        whenever(validator.validate(anyEntity(), any())).thenReturn(Result())

        val result = facade.add(entity)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.events()).isEmpty()
        }

        verify(service).add(domain)
        verify(mapper).map(entity)
        verify(validator).validate(entity, ValidationType.NEW, ValidationType.DEEP)
        verifyNoMoreInteractions(service, mapper, validator)
    }

    /**
     * Test method for [MovableParentFacade.add] with invalid data.
     */
    @Test
    fun add_InvalidData() {
        val entity = newEntity(Integer.MAX_VALUE)

        whenever(validator.validate(anyEntity(), any())).thenReturn(INVALID_DATA_RESULT)

        val result = facade.add(entity)

        assertThat(result).isEqualTo(INVALID_DATA_RESULT)

        verify(validator).validate(entity, ValidationType.NEW, ValidationType.DEEP)
        verifyNoMoreInteractions(validator)
        verifyZeroInteractions(service, mapper)
    }

    /**
     * Test method for [MovableParentFacade.update].
     */
    @Test
    fun update() {
        val entity = newEntity(1)
        val domain = newDomain(1)

        initUpdateMock(domain)

        val result = facade.update(entity)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.events()).isEmpty()
        }

        verifyUpdateMock(entity, domain)
        verifyNoMoreInteractions(service, mapper, validator)
    }

    /**
     * Test method for [MovableParentFacade.update] with invalid data.
     */
    @Test
    fun update_InvalidData() {
        val entity = newEntity(Integer.MAX_VALUE)

        whenever(validator.validate(anyEntity(), any())).thenReturn(INVALID_DATA_RESULT)

        val result = facade.update(entity)

        assertThat(result).isEqualTo(INVALID_DATA_RESULT)

        verify(validator).validate(entity, ValidationType.UPDATE, ValidationType.EXISTS, ValidationType.DEEP)
        verifyNoMoreInteractions(validator)
        verifyZeroInteractions(service, mapper)
    }

    /**
     * Test method for [MovableParentFacade.remove].
     */
    @Test
    fun remove() {
        val entity = newEntity(1)
        val domain = newDomain(1)

        whenever(service.get(any())).thenReturn(domain)
        whenever(validator.validate(anyEntity(), any())).thenReturn(Result())

        val result = facade.remove(entity)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.events()).isEmpty()
        }

        verify(service).get(1)
        verify(service).remove(domain)
        verify(validator).validate(entity, ValidationType.EXISTS)
        verifyNoMoreInteractions(service, validator)
        verifyZeroInteractions(mapper)
    }

    /**
     * Test method for [MovableParentFacade.remove] with invalid data.
     */
    @Test
    fun remove_InvalidData() {
        val entity = newEntity(Integer.MAX_VALUE)

        whenever(validator.validate(anyEntity(), any())).thenReturn(INVALID_DATA_RESULT)

        val result = facade.remove(entity)

        assertThat(result).isEqualTo(INVALID_DATA_RESULT)

        verify(validator).validate(entity, ValidationType.EXISTS)
        verifyNoMoreInteractions(validator)
        verifyZeroInteractions(service, mapper)
    }

    /**
     * Test method for [MovableParentFacade.duplicate].
     */
    @Test
    fun duplicate() {
        val entity = newEntity(1)
        val domain = newDomain(1)

        whenever(service.get(any())).thenReturn(domain)
        whenever(validator.validate(anyEntity(), any())).thenReturn(Result())

        val result = facade.duplicate(entity)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.events()).isEmpty()
        }

        verify(service).get(1)
        verify(service).duplicate(domain)
        verify(validator).validate(entity, ValidationType.EXISTS)
        verifyNoMoreInteractions(service, validator)
        verifyZeroInteractions(mapper)
    }

    /**
     * Test method for [MovableParentFacade.duplicate] with invalid data.
     */
    @Test
    fun duplicate_InvalidData() {
        val entity = newEntity(Integer.MAX_VALUE)

        whenever(validator.validate(anyEntity(), any())).thenReturn(INVALID_DATA_RESULT)

        val result = facade.duplicate(entity)

        assertThat(result).isEqualTo(INVALID_DATA_RESULT)

        verify(validator).validate(entity, ValidationType.EXISTS)
        verifyNoMoreInteractions(validator)
        verifyZeroInteractions(service, mapper)
    }

    /**
     * Test method for [MovableParentFacade.moveUp].
     */
    @Test
    fun moveUp() {
        val entity = newEntity(1)
        val domain = newDomain(1)

        whenever(service.get(any())).thenReturn(domain)
        whenever(validator.validate(anyEntity(), any())).thenReturn(Result())

        val result = facade.moveUp(entity)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.events()).isEmpty()
        }

        verify(service).get(1)
        verify(service).moveUp(domain)
        verify(validator).validate(entity, ValidationType.EXISTS, ValidationType.UP)
        verifyNoMoreInteractions(service, validator)
        verifyZeroInteractions(mapper)
    }

    /**
     * Test method for [MovableParentFacade.moveUp] with invalid data.
     */
    @Test
    fun moveUp_InvalidData() {
        val entity = newEntity(Integer.MAX_VALUE)

        whenever(validator.validate(anyEntity(), any())).thenReturn(INVALID_DATA_RESULT)

        val result = facade.moveUp(entity)

        assertThat(result).isEqualTo(INVALID_DATA_RESULT)

        verify(validator).validate(entity, ValidationType.EXISTS, ValidationType.UP)
        verifyNoMoreInteractions(validator)
        verifyZeroInteractions(service, mapper)
    }

    /**
     * Test method for [MovableParentFacade.moveDown].
     */
    @Test
    fun moveDown() {
        val entity = newEntity(1)
        val domain = newDomain(1)

        whenever(service.get(any())).thenReturn(domain)
        whenever(validator.validate(anyEntity(), any())).thenReturn(Result())

        val result = facade.moveDown(entity)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.events()).isEmpty()
        }

        verify(service).get(1)
        verify(service).moveDown(domain)
        verify(validator).validate(entity, ValidationType.EXISTS, ValidationType.DOWN)
        verifyNoMoreInteractions(service, validator)
        verifyZeroInteractions(mapper)
    }

    /**
     * Test method for [MovableParentFacade.moveDown] with invalid data.
     */
    @Test
    fun moveDown_InvalidData() {
        val entity = newEntity(Integer.MAX_VALUE)

        whenever(validator.validate(anyEntity(), any())).thenReturn(INVALID_DATA_RESULT)

        val result = facade.moveDown(entity)

        assertThat(result).isEqualTo(INVALID_DATA_RESULT)

        verify(validator).validate(entity, ValidationType.EXISTS, ValidationType.DOWN)
        verifyNoMoreInteractions(validator)
        verifyZeroInteractions(service, mapper)
    }

    /**
     * Test method for [MovableParentFacade.updatePositions].
     */
    @Test
    fun updatePositions() {
        val result = facade.updatePositions()

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.events()).isEmpty()
        }

        verify(service).updatePositions()
        verifyNoMoreInteractions(service)
        verifyZeroInteractions(mapper, validator)
    }

    /**
     * Initializes mock for update.
     *
     * @param domain domain
     */
    protected open fun initUpdateMock(domain: U) {
        whenever(mapper.map(anyEntity())).thenReturn(domain)
        whenever(validator.validate(anyEntity(), any())).thenReturn(Result())
    }

    /**
     * Verifies mock for update.
     *
     * @param entity entity
     * @param domain domain
     */
    protected open fun verifyUpdateMock(entity: T, domain: U) {
        verify(service).update(domain)
        verify(mapper).map(entity)
        verify(validator).validate(entity, ValidationType.UPDATE, ValidationType.EXISTS, ValidationType.DEEP)
    }

    /**
     * Returns facade for movable data for parent data.
     *
     * @return facade for movable data for parent data
     */
    protected abstract fun getFacade(): MovableParentFacade<T>

    /**
     * Returns entity.
     *
     * @param id ID
     * @return entity
     */
    protected abstract fun newEntity(id: Int?): T

    /**
     * Returns domain.
     *
     * @param id ID
     * @return domain
     */
    protected abstract fun newDomain(id: Int?): U

    /**
     * Returns any mock for entity.
     *
     * @return any mock for entity
     */
    protected abstract fun anyEntity(): T

    /**
     * Returns ny mock for domain.
     *
     * @return ny mock for domain
     */
    protected abstract fun anyDomain(): U

}
